### PR TITLE
[spaceship] fix creation App Store Connect API authorization token

### LINKED ISSUE #19578: upload_to_testflight  Authentication credentials are missing or invalid.
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

Uploading testflight reports an error when publishing. This happens only occasionally, but it is fatal for the CI process. It is often encountered that the testflight package is not automatically released. If this problem cannot be completely resolved, it is recommended to automatically retry.

##### Command executed

upload_to_testflight

##### Complete output when running fastlane, including the stack trace and command used

<details>
<summary>upload_to_testflight output </summary>
  <pre> 
1174 [10:23:20]: --- Step: upload_to_testflight ---
1175 [10:23:20]: ----------------------------------
1176 [10:23:20]: Creating authorization token for App Store Connect API
1177 [10:23:25]: Ready to upload new build to TestFlight (App: 999999999)...
1178 [10:23:26]: Going to upload updated app to App Store Connect
1179 [10:23:26]: This might take a few minutes. Please don't interrupt the script.
1180 [10:25:54]: iTunes Transporter successfully finished its job
1181 [10:25:54]: --------------------------------------------------------------------
1182 [10:25:54]: Successfully uploaded package to App Store Connect. It might take a few minutes until it's visible online.
1183 [10:25:54]: --------------------------------------------------------------------
1184 [10:25:54]: Successfully uploaded the new binary to App Store Connect
1185 [10:25:54]: If you want to skip waiting for the processing to be finished, use the `skip_waiting_for_build_processing` option
1186 [10:25:54]: Note that if `skip_waiting_for_build_processing` is used but a `changelog` is supplied, this process will wait for the build to appear on AppStoreConnect, update the changelog and then skip the remaining of the processing steps.
1187 [10:25:54]: Waiting for processing on... app_id: 999999999, app_version: 3.3.0, build_version: 4125, platform: IOS
1188 [10:25:56]: Read more information on why this build isn't showing up yet - https://github.com/fastlane/fastlane/issues/14997
1189 [10:25:56]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
1190 [10:26:29]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
1191 [10:27:01]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
1192 [10:27:34]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
1193 [10:28:06]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
1194 [10:28:39]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
1195 [10:29:11]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
1196 [10:29:44]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
1197 [10:30:16]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
1198 [10:30:49]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
1199 [10:31:24]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1200 [10:31:57]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1201 [10:32:30]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1202 [10:33:04]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1203 [10:33:38]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1204 [10:34:11]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1205 [10:34:45]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1206 [10:35:18]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1207 [10:35:51]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1208 [10:36:25]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1209 [10:36:58]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1210 [10:37:30]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1211 [10:38:03]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1212 [10:38:35]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1213 [10:39:08]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1214 [10:39:41]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1215 [10:40:15]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1216 [10:40:48]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1217 [10:41:21]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1218 [10:41:54]: Waiting for App Store Connect to finish processing the new build (3.3.0 - 4125) for IOS
1219 +------------------+---------------------------------------------+
1220 |                          Lane Context                          |
1221 +------------------+---------------------------------------------+
1222 | DEFAULT_PLATFORM | ios                                         |
1223 | TARGET_NAME      | {"com.example.bundleid"=>"Runner"} |
1224 | VERSION_NUMBER   | 3.3.0                                       |
1225 | PLATFORM_NAME    | ios                                         |
1226 | LANE_NAME        | ios testFlight                              |
1227 | OUTPUT_NAME      | Runner_3.3.0_4125_11.05                     |
1228 +------------------+---------------------------------------------+
1229 [10:42:25]: Authentication credentials are missing or invalid. - Provide a properly configured and signed bearer token, and make sure that it has not expired. Learn more about Generating Tokens for API Requests https://developer.apple.com/go/?id=api-generating-tokens
1230 +------+---------------------------+-------------+
1231 |                fastlane summary                |
1232 +------+---------------------------+-------------+
1233 | Step | Action                    | Time (in s) |
1234 +------+---------------------------+-------------+
1235 | 1    | default_platform          | 0           |
1236 | 2    | get_target_name           | 0           |
1237 | 3    | get_version_number        | 0           |
1238 | 4    | import_from_git           | 0           |
1239 | 5    | default_platform          | 0           |
1240 | 6    | app_store_connect_api_key | 0           |
1241 | 💥   | upload_to_testflight      | 1144        |
1242 +------+---------------------------+-------------+
1243 [10:42:25]: fastlane finished with errors
1244 Looking for related GitHub issues on fastlane/fastlane...
1245 [!] The request could not be completed because:
1246 	Authentication credentials are missing or invalid. - Provide a properly configured and signed bearer token, and make sure that it has not expired. Learn more about Generating Tokens for API Requests https://developer.apple.com/go/?id=api-generating-tokens
</pre>
</details>

### Environment

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                   |
| --------------------------- | ------------------------------------------------------- |
| OS                          | 10.15.7                                                 |
| Ruby                        | 3.0.2                                                   |
| Bundler?                    | false                                                   |
| Git                         | git version 2.24.3 (Apple Git-128)                      |
| Installation Source         | /usr/local/Cellar/fastlane/2.197.0/libexec/bin/fastlane |
| Host                        | Mac OS X 10.15.7 (19H1030)                              |
| Ruby Lib Dir                | /usr/local/Cellar/ruby/3.0.2_1/lib                      |
| OpenSSL Version             | OpenSSL 1.1.1l  24 Aug 2021                             |
| Is contained                | false                                                   |
| Is homebrew                 | true                                                    |
| Is installed via Fabric.app | false                                                   |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/             |
| Xcode Version               | 12.2                                                    |
| Swift Version               | 5.3.1                                                   |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | zh_CN.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

**No Fastfile found**

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.197.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                             | Version      |
| ------------------------------- | ------------ |
| did_you_mean                    | 1.5.0        |
| atomos                          | 0.1.3        |
| rexml                           | 3.2.5        |
| CFPropertyList                  | 3.0.4        |
| claide                          | 1.0.3        |
| colored2                        | 3.1.2        |
| nanaimo                         | 0.3.0        |
| xcodeproj                       | 1.21.0       |
| rouge                           | 2.0.7        |
| xcpretty                        | 0.3.0        |
| terminal-notifier               | 2.0.0        |
| unicode-display_width           | 1.8.0        |
| terminal-table                  | 1.8.0        |
| plist                           | 3.6.0        |
| public_suffix                   | 4.0.6        |
| addressable                     | 2.8.0        |
| multipart-post                  | 2.0.0        |
| word_wrap                       | 1.0.0        |
| optparse                        | 0.1.1        |
| tty-screen                      | 0.8.1        |
| tty-cursor                      | 0.7.1        |
| tty-spinner                     | 0.9.3        |
| artifactory                     | 3.0.15       |
| babosa                          | 1.0.4        |
| colored                         | 1.2          |
| highline                        | 2.0.3        |
| commander                       | 4.6.0        |
| excon                           | 0.87.0       |
| faraday-em_http                 | 1.0.0        |
| faraday-em_synchrony            | 1.0.0        |
| faraday-excon                   | 1.1.0        |
| faraday-httpclient              | 1.0.1        |
| faraday-net_http                | 1.0.1        |
| faraday-net_http_persistent     | 1.2.0        |
| faraday-patron                  | 1.0.0        |
| faraday-rack                    | 1.0.0        |
| ruby2_keywords                  | 0.0.5        |
| faraday                         | 1.8.0        |
| unf_ext                         | 0.0.8        |
| unf                             | 0.1.4        |
| domain_name                     | 0.5.20190701 |
| http-cookie                     | 1.0.4        |
| faraday-cookie_jar              | 0.0.7        |
| faraday_middleware              | 1.2.0        |
| fastimage                       | 2.2.5        |
| gh_inspector                    | 1.1.3        |
| json                            | 2.5.1        |
| mini_magick                     | 4.11.0       |
| naturally                       | 2.2.1        |
| rubyzip                         | 2.3.2        |
| security                        | 0.1.3        |
| xcpretty-travis-formatter       | 1.0.1        |
| dotenv                          | 2.7.6        |
| bundler                         | 2.2.22       |
| simctl                          | 1.6.8        |
| jwt                             | 2.3.0        |
| uber                            | 0.1.0        |
| declarative                     | 0.0.20       |
| trailblazer-option              | 0.1.1        |
| representable                   | 3.1.1        |
| retriable                       | 3.1.2        |
| mini_mime                       | 1.1.2        |
| memoist                         | 0.16.2       |
| multi_json                      | 1.15.0       |
| os                              | 1.1.1        |
| signet                          | 0.16.0       |
| googleauth                      | 1.0.0        |
| httpclient                      | 2.8.3        |
| webrick                         | 1.7.0        |
| google-apis-core                | 0.4.1        |
| google-apis-playcustomapp_v1    | 0.5.0        |
| google-apis-androidpublisher_v3 | 0.12.0       |
| google-cloud-env                | 1.5.0        |
| google-cloud-errors             | 1.2.0        |
| google-cloud-core               | 1.6.0        |
| google-apis-iamcredentials_v1   | 0.7.0        |
| google-apis-storage_v1          | 0.8.0        |
| rake                            | 13.0.6       |
| digest-crc                      | 0.6.4        |
| google-cloud-storage            | 1.34.1       |
| emoji_regex                     | 3.2.3        |
| jmespath                        | 1.4.0        |
| aws-partitions                  | 1.518.0      |
| aws-eventstream                 | 1.2.0        |
| aws-sigv4                       | 1.4.0        |
| aws-sdk-core                    | 3.121.3      |
| aws-sdk-kms                     | 1.50.0       |
| aws-sdk-s3                      | 1.104.0      |
| tsort                           | 0.1.0        |
| uri                             | 0.10.1       |
| set                             | 1.0.1        |
| forwardable                     | 1.3.2        |
| logger                          | 1.4.3        |
| pathname                        | 0.1.0        |
| shellwords                      | 0.1.0        |
| cgi                             | 0.2.0        |
| date                            | 3.1.0        |
| timeout                         | 0.1.1        |
| stringio                        | 3.0.0        |
| openssl                         | 2.2.0        |
| io-nonblock                     | 0.1.0        |
| ipaddr                          | 1.2.2        |
| io-wait                         | 0.1.0        |
| zlib                            | 1.1.0        |
| resolv                          | 0.2.0        |
| securerandom                    | 0.1.0        |
| digest                          | 3.0.0        |
| time                            | 0.1.0        |
| open-uri                        | 0.1.0        |
| mutex_m                         | 0.1.1        |
| net-protocol                    | 0.1.0        |
| ostruct                         | 0.3.1        |
| english                         | 0.7.1        |
| erb                             | 2.2.0        |
| strscan                         | 3.0.0        |
| abbrev                          | 0.1.0        |
| io-console                      | 0.5.7        |
| tempfile                        | 0.1.1        |
| delegate                        | 0.2.0        |
| fileutils                       | 1.5.0        |
| tmpdir                          | 0.1.2        |
| base64                          | 0.1.0        |
| singleton                       | 0.1.1        |
| net-http                        | 0.1.1        |
| open3                           | 0.1.1        |
| nkf                             | 0.1.0        |
| prettyprint                     | 0.1.0        |
| pp                              | 0.1.0        |
| find                            | 0.1.0        |
| yaml                            | 0.1.1        |
| psych                           | 3.3.0        |
</details>

*generated on:* **2021-11-05**
</details>

### PR DESCRIPTION
**Pull request by nnsnodnb**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Resolves #19578

I receive often under message.

> Authentication credentials are missing or invalid. - Provide a properly configured and signed bearer token, and make sure that it has not expired. Learn more about Generating Tokens for API Requests https://developer.apple.com/go/?id=api-generating-tokens

My `Fastfile` is following.

```ruby
default_platform(:ios)

platform :ios do
  lane :development_certificates do
    api_key_path = "./app-store-connect-api-key.json"
    match(
      api_key_path: api_key_path,
      type: "development",
      app_identifier: ["com.example.App"],
      git_url: "git@github.com:nnsnodnb/fastlane-match-certificates.git",
      git_branch: "main",
      shallow_clone: true,
      clone_branch_directly: true,
      force_for_new_devices: true,
    )
  end
end
```

and I executed.

```shell
$ fastlane ios development_certificates
```

### Description

I have made a correction to follow the JWT generated header and payload as indicated by Apple.

https://developer.apple.com/go/?id=api-generating-tokens

1. Add `typ` in header.
2. Add `iat` in payload.

### Testing Steps

Perform a non-read_only match using the App Store Connect API key.

### CODE REVIEW

**Comment by nnsnodnb** [comment]
I signed it!

**Review by ghost — APPROVED** [review]

**Review by rogerluan — COMMENTED** [review]
I don't know enough to review this PR 🙇 Friendly ping @joshdholtz, this is affecting many people (which were just masked by the fact that we released "auto-retries" in the last release) - me included 🙈

**Review by joshdholtz — APPROVED** [review]
Oh my 😱 I cannot believe that we didn't have that! Thank you so much for catching and fixing ❤️❤️❤️

**Comment by nnsnodnb** [comment]
Thank you for your review and merge 😄

File: spaceship/spec/connect_api/token_spec.rb [hunk 1]
```diff
@@ -215,10 +215,12 @@
       payload, header = JWT.decode(token.text, public_key, true, { algorithm: 'ES256' })
 
       expect(payload['iss']).to eq(issuer_id)
+      expect(payload['iat']).to eq(Time.now.to_i)
```
Comment (juan-vg) [inline spaceship/spec/connect_api/token_spec.rb:218]: @nnsnodnb Excuse me if I'm asking a silly thing, but I'm wondering if Time.now is going to be exactly the same value as the one set in the token creation time. Is not there any chance to get a race condition here?

**Pull request merged**